http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/RunnableRequest.java

```diff
@@ -74,34 +80,41 @@ public void run() {
      */
     private void initializePerRequestHandlers() {
         // Remove any existing handlers from the pipeline from the previous request.
-        removeIfExists(HttpStreamsClientHandler.class);
-        removeIfExists(ResponseHandler.class);
+        removeIfExists(HttpStreamsClientHandler.class, ResponseHandler.class,
+                       ReadTimeoutHandler.class, WriteTimeoutHandler.class,
+                       HandlerPublisher.class, HandlerSubscriber.class);
 
+        // TODO make these configurable
+        channel.pipeline().addFirst(new WriteTimeoutHandler(50));
+        channel.pipeline().addFirst(new ReadTimeoutHandler(50));
         channel.pipeline().addLast(new HttpStreamsClientHandler());
         channel.pipeline().addLast(new ResponseHandler());
     }
 
     /**
      * Removes the handler from the pipeline if present.
      *
-     * @param handler Handler to remove, identified by class.
+     * @param handlers Handlers to remove, identified by class.
      */
-    private void removeIfExists(Class<? extends ChannelHandler> handler) {
-        if (channel.pipeline().get(handler) != null) {
-            channel.pipeline().remove(handler);
+    @SafeVarargs
+    private final void removeIfExists(Class<? extends ChannelHandler>... handlers) {
+        for (Class<? extends ChannelHandler> handler : handlers) {
+            if (channel.pipeline().get(handler) != null) {
+                channel.pipeline().remove(handler);
+            }
         }
     }
 
     @Override
     public void abort() {
         if (channel != null) {
-            channel.disconnect().addListener(ignored -> context.channelPool().release(channel));
+            channel.close().addListener(ignored -> context.channelPool().release(channel));
```

Comment:
a closeAndRelease method would be good here too.

Reply:
Refactored into `closeAndRelease`